### PR TITLE
Models and all the possible errors. Will add more in future

### PR DESCRIPTION
**Pull request by AA1999**

### CODE REVIEW

**Review by sebkuip — CHANGES_REQUESTED** [review]
There is also a random binary file, if it could be removed:
discord/types/_disthon_discord_types_snowflake.py1.stats

File: discord/client/client.py [hunk 1]
```diff
@@ -1,94 +1,92 @@
...
-            await self.handle_event(global_message)
-        try:
-
-            for coro in self.events[event]:
-                await coro(msg)
-        except KeyError:
-            return
+import inspect
+from copy import deepcopy
+from discord.api import intents as intent, handler, websocket
```
[the remark this answers is not in the document]
Comment (AA1999) [reply discord/client/client.py:3]: This must be an older update of your code

File: discord/errors/discordexception.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import traceback
```
[the remark this answers is not in the document]
Comment (AA1999) [reply discord/errors/discordexception.py:1]: I'll work on getting them all in one file

File: discord/abc/abstarctuser.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from discord.abc.discordobject import DiscordObject
+from discord.types.avatar import Avatar
+
+
+class AbstractUser(DiscordObject):
```
Comment (sebkuip) [inline discord/abc/abstarctuser.py:5]: This file still isn't deleted. This is the one with the typo in the name

File: discord/abc/discordobject.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+from abc import ABCMeta
```
Comment (sebkuip) [inline discord/abc/discordobject.py:1]: Where is this ABCMeta defined?

File: discord/abc/discordobject.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        return self._created_at
+
+    def __eq__(self, other):
+        return isinstance(other, DiscordObject) and self.id == other.id
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return self.id.id >> 22
```
Comment (sebkuip) [inline discord/abc/discordobject.py:28]: self.id.id looks extremely off and wrong.

File: discord/activity/activity.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from discord.activity.presenseassets import PresenceAssets
+from discord.types.snowflake import Snowflake
+from discord.activity.baseactivity import BaseActivity
+
+
+class Activity(BaseActivity):
```
Comment (sebkuip) [inline discord/activity/activity.py:6]: `__slots__` is missing here

File: discord/activity/baseactivity.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+from datetime import datetime
+
+from discord.types.activitypayload import ActivityPayload
+
+
+class BaseActivity:
+    _created_at: Optional[datetime]
+    __slots__ = ('_created_at',)
+    
+    def __init(self, **kwargs):
```
Comment (sebkuip) [inline discord/activity/baseactivity.py:11]: forgot the right side `__` for the init

File: discord/activity/presenseassets.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class PresenceAssets(dict[Snowflake, str]):
+    largeimage: Snowflake
+    largetext: str
+    smallimage: Snowflake
+    smalltext: str
+
+    def __init__(self, activity: Activity, assets: RawActivityAssets):
+        pass
```
Comment (sebkuip) [inline discord/activity/presenseassets.py:15]: Either fill this in or add a notimplementederror

File: discord/activity/presenseassets.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
+from __future__ import annotations
+
+from discord.activity.rawactivityassets import RawActivityAssets
+from discord.types.snowflake import Snowflake
+from discord.activity.activity import Activity
+
+
+class PresenceAssets(dict[Snowflake, str]):
```
Comment (sebkuip) [inline discord/activity/presenseassets.py:8]: `__slots__` is missing

File: discord/activity/rawactivityassets.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+from typing import TypedDict
+
+from discord.types.snowflake import Snowflake
+
+
+class RawActivityAssets(TypedDict):
```
Comment (sebkuip) [inline discord/activity/rawactivityassets.py:6]: `__slots__` is missing

File: discord/api/handler.py [hunk 11]
```diff
@@ -1,86 +1,88 @@
...
+
+class Handler:
+    def __init__(self):
+        self.base_url: str = 'https://discord.com/api/v9/'
+        self.user_agent: str = "Disthon test library V0.0.1b"
+
+    async def request(self, method: str, dest: str, *, headers: typing.Optional[dict] = None,
+                      data: typing.Optional[dict] = None) -> typing.Union[str, dict]:
+        async with self.__session.request(method, self.base_url + dest, headers=headers, json=data) as r:
+            if not 200 <= r.status < 300 and r.status == 401:
```
Comment (sebkuip) [inline discord/api/handler.py:15]: The nested if was to be able to add more error codes in the future for different types of errors

File: discord/api/websocket.py [hunk 12]
```diff
@@ -1,141 +1,143 @@
...
+    async def identify(self) -> None:
+        """Sends the IDENTIFY packet"""
+        payload = {
+            'op': self.IDENTIFY,
+            'd': {
+                'token': self.token,
+                'intents': self.client.intents.value,
+                'properties': {
+                    '$os': sys.platform,
+                    '$browser': 'discord',
```
Comment (sebkuip) [inline discord/api/websocket.py:123]: We're not discord accessing itself. We're disthon the library

File: discord/color/color.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+    _value: int
+
+    @staticmethod
+    def _is_16_bit(color: Union[int, str]) -> bool:
+        if isinstance(color, str):
+            if color.startswith('#'):
+                if not color[1].isdigit():
+                    return False
+                for c in color[1]:
```
Comment (sebkuip) [inline discord/color/color.py:18]: Slicing forgotten

File: discord/color/color.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+                    try:
+                        a = int(c, 16)
+                    except ValueError:
+                        return False
+                return True
+            if color.startswith('0x'):
+                if not color[2].isdigit():
+                    return False
+                for c in color[2]:
```
Comment (sebkuip) [inline discord/color/color.py:27]: Same as above. Did you forget slicing?

File: discord/color/color.py [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+                if not color[1].isdigit():
+                    return False
+                for c in color[1]:
+                    try:
+                        a = int(c, 16)
+                    except ValueError:
+                        return False
+                return True
+            if color.startswith('0x'):
+                if not color[2].isdigit():
```
Comment (sebkuip) [inline discord/color/color.py:25]: Slicing was forgotten here

File: discord/color/color.py [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+class Color:
+    __slots__ = '_value'
+
+    _value: int
+
+    @staticmethod
+    def _is_16_bit(color: Union[int, str]) -> bool:
+        if isinstance(color, str):
+            if color.startswith('#'):
+                if not color[1].isdigit():
```
Comment (sebkuip) [inline discord/color/color.py:16]: Why the `[1]` index here? Did you forget a `:` for slicing?

File: discord/errors/discordexception.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sebkuip) [reply discord/errors/discordexception.py:1]: Whilst you did move them all to 1 file, you forgot to delete the separate files

File: discord/guild/guild.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+from discord.abc.discordobject import DiscordObject
+from discord.channels.guildchannel import GuildChannel
+from discord.member.member import Member
+from discord.role.role import Role
+from discord.types.guildpayload import GuildPayload
+from discord.types.snowflake import Snowflake
+from discord.user.user import User
+
+
+class BanEntry(NamedTuple):
```
Comment (sebkuip) [inline discord/guild/guild.py:15]: Maybe add `__slots__`?

File: discord/guild/guild.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    user: User
+    reason: Optional[str]
+
+
+class GuildLimit(NamedTuple):
```
Comment (sebkuip) [inline discord/guild/guild.py:20]: `__slots__` here too?

File: discord/guild/guild.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    roles: int
+    categories: int
+    bitrate: int
+    stickers: int
+
+
+class Guild(DiscordObject):
+    __slots__ = (
+        'region'
+        'owner.id'
```
Comment (sebkuip) [inline discord/guild/guild.py:33]: Did you mean `owner_id`?

File: discord/guild/guild.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        'mfa.level'
```
Comment (sebkuip) [inline discord/guild/guild.py:34]: Did you mean `moderation_level`? MFA is just a True/False for guilds

File: discord/guild/guild.py [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+        '_vanity'
+        '_banner'
+    )
+
+    _roles: set[Role]
+    me: Member
+    owner_id: Snowflake
+
+    def __init__(self, data: GuildPayload):
+        self._members: dict[Snowflake, Member] = {}
```
Comment (sebkuip) [inline discord/guild/guild.py:48]: I assume these are for the cache and not directly for the guild? Are they misplaced?

File: discord/interaction/interaction.py [hunk 22]
```diff
@@ -0,0 +1,2 @@
+class Interaction:
+    pass
```
Comment (sebkuip) [inline discord/interaction/interaction.py:2]: Possibly make this `notimplementederror`?

File: discord/internal/cache.py [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+from typing import Any, OrderedDict
+from discord.types.snowflake import Snowflake
+from discord.user.user import User
+from discord.member.member import Member
+from discord.message.message import Message
+from discord.role.role import Role
+from discord.guild.guild import Guild
+from discord.api.handler import Handler
+
```
Comment (sebkuip) [inline discord/internal/cache.py:11]: Inside this folder there is a random .stats file. Not sure if that's supposed to be there?

File: discord/internal/cache.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+
+class LFUCache:
+    __slots__ = ('_max_size', '_cache', '_frequency')
+
+    _max_size: int
+    _cache: OrderedDict[Snowflake, Any]
+    _frequency: dict[Snowflake, int]
+    _handler: Handler
```
Comment (sebkuip) [inline discord/internal/cache.py:19]: Why is handler not in slots?

File: discord/internal/cache.py [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+            return self._cache[key]
+        raise KeyError
+
+    def __delitem__(self, key: Snowflake):
+        del self._cache[key]
+        del self._frequency[key]
+
+
+class UserCache(LFUCache):
+    _cache: dict[Snowflake, User]
```
Comment (sebkuip) [inline discord/internal/cache.py:70]: This is already set in the baseclass as a slot. This is redundant. Please also adress this in every other type of cache thanks

File: discord/member/member.py [hunk 26]
```diff
@@ -0,0 +1,10 @@
+from __future__ import annotations
+
+from discord.role.role import Role
+from discord.user.user import User
+
+
+class Member(User):
+    top_role: Role
+    roles: set[Role]
+
```
Comment (sebkuip) [inline discord/member/member.py:10]: Add `__slots__` and a guild

File: discord/role/role.py [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+
+    def __repr__(self):
+        return f'Role {self._name} with id {self._id}'
+
+    def __eq__(self, other):
+        return isinstance(other, Role) and self.id == other.id and self.guild == other.guild
+
+    def __lt__(self, other):
+        if not isinstance(other, Role):
+            raise NotImplementedError
```
Comment (sebkuip) [inline discord/role/role.py:81]: I'd say if we're not comparing roles we should raise a typeerror

File: discord/types/snowflake.py [hunk 28]
```diff
@@ -0,0 +1,46 @@
+from typing import Union, Any
+
```
Comment (sebkuip) [inline discord/types/snowflake.py:2]: There is a random .stats file. Could you remove it?

File: discord/types/banner.py [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+    def __init__(self, url: str, color: Optional[Color]):
+        super().__init__(url)
+        self._color = color
+
+    @property
+    def color(self):
+        return self._color
+
+    @property
+    def colour(self):
```
Comment (sebkuip) [inline discord/types/banner.py:20]: Why not make this a `colour = color` since it's just an alias?

File: discord/types/rolepayload.py [hunk 30]
```diff
@@ -0,0 +1,22 @@
+from typing import TypedDict
+
+from discord.color.color import Color
+from discord.types.snowflake import Snowflake
+
+
+class RoleTagsPayload(TypedDict, total=False):
+    bot_id: Snowflake
```
Comment (sebkuip) [inline discord/types/rolepayload.py:8]: Why not `__slots__`?

File: discord/types/userpayload.py [hunk 31]
```diff
@@ -0,0 +1,22 @@
+from discord.types.enums.locale import Locale
+from typing import TypedDict, Optional
+
+from discord.types.enums.userflags import UserFlags
+from discord.types.snowflake import Snowflake
+
+
+class UserPayload(TypedDict):
+    id: Snowflake
```
Comment (sebkuip) [inline discord/types/userpayload.py:9]: Please add `__slots__`

File: discord/user/baseuser.py [hunk 32]
```diff
@@ -0,0 +1,119 @@
...
+from discord.abc.abstractuser import AbstractUser
+from discord.color.color import Color
+from discord.internal.cache import GuildCache, UserCache
+from discord.message.message import Message
+from discord.types.avatar import Avatar
+from discord.types.banner import Banner
+from discord.types.enums.defaultavatar import DefaultAvatar
+from discord.types.enums.userflags import UserFlags
+from discord.types.userpayload import UserPayload
+
```
Comment (sebkuip) [inline discord/user/baseuser.py:13]: Random .stats file here too

**Pull request closed without merging**